search/dijkstra.go

```diff
@@ -0,0 +1,187 @@
+	if !g.NodeExists(u) {
+		return Shortest{from: u}
+	}
+	var (
+		from   = g.Neighbors
+		edgeTo func(graph.Node, graph.Node) graph.Edge
+	)
+	switch g := g.(type) {
+	case graph.DirectedGraph:
+		from = g.Successors
+		edgeTo = g.EdgeTo
+	default:
+		edgeTo = g.EdgeBetween
+	}
+	if weight == nil {
+		if g, ok := g.(graph.Coster); ok {
+			weight = g.Cost
+		} else {
+			weight = UniformCost
+		}
+	}
+
+	nodes := g.NodeList()
+
+	indexOf := make(map[int]int, len(nodes))
+	for i, n := range nodes {
+		indexOf[n.ID()] = i
+	}
+
+	path := Shortest{
+		from: u,
+
+		nodes:   nodes,
+		indexOf: indexOf,
+
+		dist: make([]float64, len(nodes)),
+		next: make([]int, len(nodes)),
+	}
+	for i := range path.dist {
+		path.dist[i] = math.Inf(1)
+	}
+
+	// Dijkstra's algorithm here is implemented essentially as
+	// described in Function B.2 in figure 6 of UTCS Technical
+	// Report TR-07-54.
+	//
+	// http://www.cs.utexas.edu/ftp/techreports/tr07-54.pdf
+	Q := priorityQueue{{node: u, dist: 0}}
+	for Q.Len() != 0 {
+		mid := heap.Pop(&Q).(distanceNode)
+		k := path.indexOf[mid.node.ID()]
+		if mid.dist < path.dist[k] {
+			path.dist[k] = mid.dist
+		}
+		for _, v := range from(mid.node) {
+			j := path.indexOf[v.ID()]
+			w := weight(edgeTo(mid.node, v))
+			if w < 0 {
+				panic("dijkstra: negative edge weight")
+			}
+			joint := path.dist[k] + w
+			if joint < path.dist[j] {
+				heap.Push(&Q, distanceNode{node: v, dist: joint})
+				path.set(j, joint, k)
+			}
+		}
+	}
+
+	return path
+}
+
+// DijkstraAllPaths returns a shortest-path tree for shortest paths in the graph g.
+// If weight is nil and the graph does not implement graph.Coster, UniformCost is used.
+// DijkstraAllPaths will panic if g has a negative edge weight.
+func DijkstraAllPaths(g graph.Graph, weight graph.CostFunc) (paths ShortestPaths) {
+	var (
+		from   = g.Neighbors
+		edgeTo func(graph.Node, graph.Node) graph.Edge
+	)
+	switch g := g.(type) {
+	case graph.DirectedGraph:
+		from = g.Successors
+		edgeTo = g.EdgeTo
+	default:
+		edgeTo = g.EdgeBetween
+	}
+	if weight == nil {
+		if g, ok := g.(graph.Coster); ok {
+			weight = g.Cost
+		} else {
+			weight = UniformCost
+		}
+	}
+
+	nodes := g.NodeList()
+
+	indexOf := make(map[int]int, len(nodes))
+	for i, n := range nodes {
+		indexOf[n.ID()] = i
+	}
+
+	dist := make([]float64, len(nodes)*len(nodes))
+	for i := range dist {
+		dist[i] = math.Inf(1)
+	}
+	paths = ShortestPaths{
+		nodes:   nodes,
+		indexOf: indexOf,
+
+		dist:    mat64.NewDense(len(nodes), len(nodes), dist),
+		next:    make([][]int, len(nodes)*len(nodes)),
+		forward: false,
+	}
+
+	var Q priorityQueue
+	for i, u := range nodes {
+		// Dijkstra's algorithm here is implemented essentially as
+		// described in Function B.2 in figure 6 of UTCS Technical
+		// Report TR-07-54 with the addition of handling multiple
+		// co-equal paths.
+		//
+		// http://www.cs.utexas.edu/ftp/techreports/tr07-54.pdf
+
+		Q = Q[:0]
```

Comment:
The reasoning is that Dijkstra is usually presented as the SSSP case, but we are providing the APSP function. The comment above explains the provenance of the code within the loop block, as being the complete SSSP+ which has a Q <- 0. I could just have a comment here that the queue is always empty here, but this does the same thing. I would like the note, so is a comment to this effect OK with you?